File: hard/Power of Thor - Episode 2/solution.cs
using System;
using System.Linq;
using System.Collections.Generic;

class Player
{
    const int HEIGHT = 18;
    const int WIDTH = 40;
    static readonly (int, int)[] DIRS = new (int dy, int dx)[] { (-1, -1), (-1, 0), (-1, 1), (1, -1), (1, 0), (1, 1), (0, -1), (0, 1) };

    static int TY;
    static int TX;
    static int H;              // the remaining number of hammer strikes.
    static int N;              // the number of giants which are still present on the map.

    static int giantsInArea;
    static bool inDanger;
    static bool surrounded;
    static (int, int)? target;

    static bool[,] forbiddenCells = new bool[HEIGHT, WIDTH];

    static void Main(string[] args)
    {
        string[] inputs;
        inputs = Console.ReadLine().Split(' ');

        TX = int.Parse(inputs[0]);
        TY = int.Parse(inputs[1]);
        (int ly, int lx) lastPos = (TY, TX);

        while (true)
        {
            ClearStates();

            inputs = Console.ReadLine().Split(' ');
            H = int.Parse(inputs[0]);
            N = int.Parse(inputs[1]);

            var gCoords = ReadGiantCoordinates();
            var barycenter = GetBarycenter(gCoords);

            var env = CheckEnv();
            inDanger = env.inDanger;
            surrounded = env.surrounded;

            target = FindNextStep(barycenter, lastPos);

            lastPos = (TY, TX);
            // The movement or action to be carried out: WAIT STRIKE N NE E SE S SW W or N
            Console.WriteLine(MakeDecision());
        }
    }

    static string MakeDecision() => CanStrike() ? "STRIKE" : GoToTarget(target.Value.Item1, target.Value.Item2);

    static string GoToTarget(int ty, int tx)
    {
        if (TY == ty && TX == tx) return "WAIT";

        string direction = "";
        direction += TY < ty ? "S" : TY > ty ? "N" : "";
        direction += TX < tx ? "E" : TX > tx ? "W" : "";

        ApplyNewPos(ty, tx);

        return direction;
    }

    static void ApplyNewPos(int targetY, int targetX)
    {
        TY += TY < targetY ? 1 : TY > targetY ? -1 : 0;
        TX += TX < targetX ? 1 : TX > targetX ? -1 : 0;
    }

    static bool CanStrike() => giantsInArea >= ((N + H - 1) / H) || surrounded;

    static (int, int)? FindNextStep((int, int) barycenter, (int, int) lastPos)
    {
        return inDanger ? GetClosestCandidate(GetNeighbors((TY, TX)), barycenter, lastPos) : barycenter;
    }

    static (int, int)? GetClosestCandidate(List<(int, int)> candidates, (int y, int x) target, (int, int) lastPos)
    {
        double bestDist = Double.MaxValue;
        (int, int)? bestC = null;

        foreach ((int cy, int cx) in candidates)
        {
            if (forbiddenCells[cy, cx]) continue;

            double dist = Math.Pow(target.y - cy, 2) + Math.Pow(target.x - cx, 2);
            if (dist < bestDist && (cy, cx) != lastPos)
            {
                bestDist = dist;
                bestC = (cy, cx);
            }
        }
        return bestC.HasValue ? bestC.Value : null;
    }

    static (bool inDanger, bool surrounded) CheckEnv()
    {
        bool inDanger = false, surrounded = true;
        foreach (var (ny, nx) in GetNeighbors((TY, TX)))
        {
            if (forbiddenCells[ny, nx] && !inDanger)
                inDanger = true;
            if (!forbiddenCells[ny, nx] && surrounded)
                surrounded = false;
        }
        return (inDanger, surrounded);
    }

    static (int, int) GetBarycenter((int[] ys, int[] xs) gCoords)
    {
        int y = gCoords.ys.Sum() / N;
        int x = gCoords.xs.Sum() / N;

        return (y, x);
    }

    static (int[] ys, int[] xs) ReadGiantCoordinates()
    {
        (int[] ys, int[] xs) coords = (new int[N], new int[N]);

        for (int i = 0; i < N; i++)
        {
            string[] inputs = Console.ReadLine().Split(' ');
            int X = int.Parse(inputs[0]);
            int Y = int.Parse(inputs[1]);

            coords.ys[i] = Y;
            coords.xs[i] = X;

            PopulateForbiddenCells(Y, X);

            if (IsInStrikeArea(Y, X)) giantsInArea++;
        }

        return coords;
    }

    static bool IsInStrikeArea(int gy, int gx) => Math.Abs(gy - TY) <= 5 && Math.Abs(gx - TX) <= 5;

    static void PopulateForbiddenCells(int y, int x)
    {
        forbiddenCells[y, x] = true;
        foreach ((int ny, int nx) in GetNeighbors((y, x)))
        {
            forbiddenCells[ny, nx] = true;
        }
    }

    static List<(int, int)> GetNeighbors((int, int) center)
    {
        (int y, int x) = center;
        List<(int, int)> neighbors = new List<(int, int)>();

        foreach (var (dy, dx) in DIRS)
        {
            int ny = y + dy;
            int nx = x + dx;
            if (IsInBounds(ny, nx))
                neighbors.Add((ny, nx));
        }
        return neighbors;
    }

    static bool IsInBounds(int y, int x) => x >= 0 && x < WIDTH && y >= 0 && y < HEIGHT;

    static void ClearStates()
    {
        giantsInArea = 0;
        inDanger = false;
        Array.Clear(forbiddenCells, 0, forbiddenCells.Length);
    }
}

File: hard/Power of Thor - Episode 2/solution.java
import java.util.*;

class Player {
    static final int HEIGHT = 18;
    static final int WIDTH = 40;
    static final int[][] DIRS = new int[][] { { -1, -1 }, { -1, 0 }, { -1, 1 }, { 1, -1 }, { 1, 0 }, { 1, 1 },
            { 0, -1 }, { 0, 1 } };

    static Scanner in = new Scanner(System.in);;

    static int TY;
    static int TX;
    static int H; // the remaining number of hammer strikes.
    static int N; // the number of giants which are still present on the map.

    static int giantsInArea;
    static boolean inDanger;
    static boolean surrounded;
    static int[] target;

    static boolean[][] forbiddenCells = new boolean[HEIGHT][WIDTH];

    public static void main(String args[]) {
        // Scanner in = new Scanner(System.in);
        TX = in.nextInt();
        TY = in.nextInt();
        int[] lastPos = new int[] { TY, TX };

        // game loop
        while (true) {
            H = in.nextInt(); // the remaining number of hammer strikes.
            N = in.nextInt(); // the number of giants which are still present on the map.

            clearStates();

            int[][] gCoords = readGiantCoordinates();
            int[] barycenter = getBarycenter(gCoords);

            var env = checkEnv();
            inDanger = env[0];
            surrounded = env[1];

            target = findNextStep(barycenter, lastPos);

            lastPos = new int[] { TY, TX };
            // The movement or action to be carried out: WAIT STRIKE N NE E SE S SW W or N
            System.out.println(makeDecision());
        }
    }

    static String makeDecision() {
        return canStrike() ? "STRIKE" : goToTarget(target);
    }

    static String goToTarget(int[] t) {
        if (TY == t[0] && TX == t[1])
            return "WAIT";

        String direction = "";
        direction += TY < t[0] ? "S" : TY > t[0] ? "N" : "";
        direction += TX < t[1] ? "E" : TX > t[1] ? "W" : "";

        applyNewPos(t);

        return direction;
    }

    static void applyNewPos(int[] target) {
        TY += TY < target[0] ? 1 : TY > target[0] ? -1 : 0;
        TX += TX < target[1] ? 1 : TX > target[1] ? -1 : 0;
    }

    static boolean canStrike() {
        return giantsInArea >= ((N + H - 1) / H) || surrounded;
    }

    static int[] findNextStep(int[] barycenter, int[] lastPos) {
        return inDanger ? getClosestCandidate(getNeighbors(TY, TX), barycenter, lastPos) : barycenter;
    }

    static int[] getClosestCandidate(List<int[]> candidates, int[] target, int[] lastPos) {
        double bestDist = Double.MAX_VALUE;
        int[] bestC = null;

        for (int[] c : candidates) {
            if (forbiddenCells[c[0]][c[1]] || (c[0] == lastPos[0] && c[1] == lastPos[1]))
                continue;

            double dist = Math.pow(target[0] - c[0], 2) + Math.pow(target[1] - c[1], 2);
            if (dist < bestDist) {
                bestDist = dist;
                bestC = c;
            }
        }
        return bestC;
    }

    static boolean[] checkEnv() {
        boolean inDanger = false, surrounded = true;
        for (int[] n : getNeighbors(TY, TX)) {
            if (forbiddenCells[n[0]][n[1]] && !inDanger)
                inDanger = true;
            if (!forbiddenCells[n[0]][n[1]] && surrounded)
                surrounded = false;
        }
        return new boolean[] { inDanger, surrounded };
    }

    static int[] getBarycenter(int[][] gCoords) {
        int y = Arrays.stream(gCoords[0]).sum() / N;
        int x = Arrays.stream(gCoords[1]).sum() / N;

        return new int[] { y, x };
    }

    static int[][] readGiantCoordinates() {
        int[][] coords = new int[2][N];

        for (int i = 0; i < N; i++) {
            int X = in.nextInt();
            int Y = in.nextInt();

            coords[0][i] = Y;
            coords[1][i] = X;

            populateForbiddenCells(Y, X);

            if (isInStrikeArea(Y, X))
                giantsInArea++;
        }

        return coords;
    }

    static boolean isInStrikeArea(int gy, int gx) {
        return Math.abs(gy - TY) <= 5 && Math.abs(gx - TX) <= 5;
    }

    static void populateForbiddenCells(int y, int x) {
        forbiddenCells[y][x] = true;
        for (int[] n : getNeighbors(y, x)) {
            forbiddenCells[n[0]][n[1]] = true;
        }
    }

    static List<int[]> getNeighbors(int y, int x) {
        List<int[]> neighbors = new ArrayList<int[]>();

        for (int[] d : DIRS) {
            int ny = y + d[0];
            int nx = x + d[1];
            if (isInBounds(ny, nx))
                neighbors.add(new int[] { ny, nx });
        }
        return neighbors;
    }

    static boolean isInBounds(int y, int x) {
        return x >= 0 && x < WIDTH && y >= 0 && y < HEIGHT;
    }

    static void clearStates() {
        giantsInArea = 0;
        inDanger = false;
        for (int y = 0; y < forbiddenCells.length; y++) {
            Arrays.fill(forbiddenCells[y], false);
        }
    }
}